pkg/session/session.go

```diff
@@ -970,6 +972,24 @@ func (s *session) RollbackTxn(ctx context.Context) {
 	sessiontxn.GetTxnManager(s).OnTxnEnd()
 }
 
+// setLastTxnInfoBeforeTxnEnd sets the @@last_txn_info variable before commit/rollback the transaction.
+// The `LastTxnInfo` updated with a JSON string that contains start_ts, for_update_ts, etc.
+// It updated `LastTxnInfo` does not contain the `commit_ts` fields because it is unknown
+// until the commit is done (or do not need to commit for readonly or a rollback transaction).
+// The non-readonly transaction will overwrite the `LastTxnInfo` again after commit to update the `commit_ts` field.
+func (s *session) setLastTxnInfoBeforeTxnEnd() {
+	txnCtx := s.GetSessionVars().TxnCtx
+	if txnCtx.StartTS == 0 {
+		// If the txn is not active, for example, executing "SELECT 1", skip setting the last txn info.
+		return
+	}
+
+	s.GetSessionVars().LastTxnInfo = fmt.Sprintf(
+		`{"txn_scope": "%s", "start_ts": %d, "for_update_ts": %d}`,
+		txnCtx.TxnScope, txnCtx.StartTS, txnCtx.GetForUpdateTS(),
+	)
```

Comment:
Consider using json serialization instead of text formatting to prevent potential mistakes. See the original logic for setting last txn info, which is located in client-go: https://github.com/tikv/client-go/blob/2058fbc06214bce4a71986d62a6ada9e89d013ff/txnkv/transaction/txn.go#L1033-L1048

Reply:
Seems if a txn comitted or fails to commit, the CommitHook will update it: https://github.com/pingcap/tidb/blob/c71ebaf73ed9620745f327ad147f27fc06d73297/pkg/sessiontxn/isolation/base.go#L506

`setLastTxnInfoBeforeTxnEnd` in `CommitTxn` will be overridden at this time. So `setLastTxnInfoBeforeTxnEnd` is only affects some readonly txns.